src/base/par_iter.rs

```diff
@@ -0,0 +1,156 @@
+        bridge(self, consumer)
+    }
+
+    fn with_producer<CB: rayon::iter::plumbing::ProducerCallback<Self::Item>>(
+        self,
+        callback: CB,
+    ) -> CB::Output {
+        let producer = ColumnIter::new(self.mat);
+        callback.callback(producer)
+    }
+}
+
+impl<T, R: Dim, Cols: Dim, S: RawStorage<T, R, Cols>> Matrix<T, R, Cols, S>
+where
+    T: Send + Sync + Clone + Debug + PartialEq + 'static,
+    S: Sync,
+{
+    /// Iterate through the columns of the matrix in parallel using rayon.
+    pub fn par_column_iter(&self) -> ParColumnIter<'_, T, R, Cols, S> {
+        ParColumnIter::new(self)
+    }
+}
+
+/// A rayon parallel iterator through the mutable columns of a matrix
+pub struct ParColumnIterMut<
+    'a,
+    T,
+    R: Dim,
+    Cols: Dim,
+    S: RawStorage<T, R, Cols> + RawStorageMut<T, R, Cols>,
+> {
+    mat: &'a mut Matrix<T, R, Cols, S>,
+}
+
+impl<'a, T, R, Cols, S> ParColumnIterMut<'a, T, R, Cols, S>
+where
+    R: Dim,
+    Cols: Dim,
+    S: RawStorage<T, R, Cols> + RawStorageMut<T, R, Cols>,
+{
+    /// create a new parallel iterator for the given matrix
+    fn new(mat: &'a mut Matrix<T, R, Cols, S>) -> Self {
+        Self { mat }
+    }
+}
+
+impl<'a, T, R, Cols, S> ParallelIterator for ParColumnIterMut<'a, T, R, Cols, S>
+where
+    R: Dim,
+    Cols: Dim,
+    S: RawStorage<T, R, Cols> + RawStorageMut<T, R, Cols>,
+    T: Send + Sync + Debug + PartialEq + Clone + 'static,
+    S: Send + Sync,
+{
+    type Item = MatrixSliceMut<'a, T, R, U1, S::RStride, S::CStride>;
+    fn drive_unindexed<C>(self, consumer: C) -> C::Result
+    where
+        C: rayon::iter::plumbing::UnindexedConsumer<Self::Item>,
+    {
+        bridge(self, consumer)
+    }
+
+    fn opt_len(&self) -> Option<usize> {
+        Some(self.mat.ncols())
+    }
+}
+
+impl<'a, T, R, Cols, S> IndexedParallelIterator for ParColumnIterMut<'a, T, R, Cols, S>
+where
+    R: Dim,
+    Cols: Dim,
+    S: RawStorage<T, R, Cols> + RawStorageMut<T, R, Cols>,
+    T: Send + Sync + Debug + PartialEq + Clone + 'static,
+    S: Send + Sync,
+{
+    fn drive<C: rayon::iter::plumbing::Consumer<Self::Item>>(self, consumer: C) -> C::Result {
+        bridge(self, consumer)
+    }
+
+    fn len(&self) -> usize {
+        self.mat.ncols()
+    }
+
+    fn with_producer<CB: rayon::iter::plumbing::ProducerCallback<Self::Item>>(
+        self,
+        callback: CB,
+    ) -> CB::Output {
+        let producer = ColumnIterMut::new(self.mat);
+        callback.callback(producer)
+    }
+}
+
+impl<T, R: Dim, Cols: Dim, S: RawStorage<T, R, Cols> + RawStorageMut<T, R, Cols>>
+    Matrix<T, R, Cols, S>
+where
+    T: Send + Sync + Clone + Debug + PartialEq + 'static,
+    S: Sync,
+{
+    /// Mutably iterate through the columns of this matrix in parallel using rayon
+    pub fn par_column_iter_mut(&mut self) -> ParColumnIterMut<'_, T, R, Cols, S> {
+        ParColumnIterMut::new(self)
```

Comment:
I think it would be nice to include an example in the docs here for how the user can use this functionality in practice. Same goes for `par_column_iter`.

Reply:
I'll try and add some nicer documentation.

*TODO* @geo-ant

Reply:
I've added better documentation (so I hope) and examples. I've also advertized the feature on the main page of the `Matrix` struct using an additional section.